### ISSUE TITLE
Move `ido.nvim` to a separate GitHub Organization

### ISSUE THREAD

**Issue by alexanderjeurissen** [ido-nvim/ido.nvim#17]
With the modularization shaping up quite nicely in https://github.com/shoumodip/ido.nvim/pull/15
and the recent inclusion of `fzy`, All that is left to do (aside from squashing bugs) is implementing a `extension` system.

By doing so we enable other plugin authors (and ourselves) to extract meaningful `extensions` (like menus.lua) to separate plugins/repositories.

It would be great if we can create a `ido.nvim` organization that way we can gather all `ido.nvim` extensions in a single place and create a ecosystem around `ido.nvim`

**Comment by alexanderjeurissen** [comment]
@shoumodip WDYT ? 

Aready created a simple `find files` using `ripgrep` we could easily push this as a plugin in a `ido.nvim` GitHub org.

```lua
local ido = require "ido"
local main = require("ido.core.main")
local advice = require("ido.core.advice")

local set_keymap = vim.api.nvim_set_keymap

local COMMAND = 'rg --files --ignore --smart-case --hidden --follow --no-messages --ignore-file ~/.gitignore'

local find_files = {}

local function edit()
  local variables = ido.sandbox.variables
  local selected = ""

  -- Check if something matched or not
  if #variables.results > 0 then
    selected = variables.results[variables.selected][1]
  else
    -- The query is the thing we need to work with
    selected = variables.before_cursor..variables.after_cursor
  end

  main.exit()

  vim.cmd("edit " .. vim.loop.cwd() .. '/' .. selected)
end

local function prompt_title()
  local directory_name = vim.loop.cwd()
  local prompt_title =  " Find file: %s/"

  return string.format(prompt_title,
  string.gsub(vim.fn.resolve(directory_name), '^' .. os.getenv('HOME'), '~'))
end

find_files.run = function (layout)
  advice.set("exit_on_accept_selected", "overwrite", edit, false)
  advice.set("single_result_after_accept_suggestion", "overwrite", edit, false)

  main.start({
    prompt = prompt_title(),
    layout = layout,
    items = vim.fn.systemlist(COMMAND),
  })
end

return find_files
```

which works great in vertical and horizontal modes 🕺 

<img width="1792" alt="Screen Shot 2021-02-23 at 10 05 14 PM" src="https://user-images.githubusercontent.com/1220084/108908175-4cd6a380-7623-11eb-8e33-7679ed94dcc8.png">
<img width="332" alt="Screen Shot 2021-02-23 at 10 04 56 PM" src="https://user-images.githubusercontent.com/1220084/108908182-4e07d080-7623-11eb-8a77-b657ac90b398.png">

**Comment by shoumodip** [comment]
Uh, why are you using advices for the action to do after selecting a value. Shouldn't this be enough?

```lua
local file = main.start({
   prompt = prompt_title(),
   layout = layout,
   items = vim.fn.systemlist(COMMAND),
})

if #file > 0 then
   edit(file)
end
```

**Comment by shoumodip** [comment]
There is no need to include the `main` module for this functionality. `ido.start()` is an alias to the `ido.core.main.start()` function. And like I said before, there is no need to use advices for this, so the `main.exit()` function is not something you need here.

If there is some specific reason you need advices to do this, and not the regular return feature, let me know. I have never used `rg` myself so I might be saying something stupid here.

**Comment by alexanderjeurissen** [comment]
Oh wow yeah it felt odd to have to use the advices for this. 

Must have misread the docs didn't know Ido.start returns the selected file. 

That will cleanup my find files function for sure.

**Comment by shoumodip** [comment]
Nah you didn't misread it. I simply forgot to mention that `ido.start()` or `ido.core.main.start()` returns the selected item or the query based on the `strict_match` option in the documentation.

**Comment by shoumodip** [comment]
After creating the `ido.nvim` organization should we move this repo into that, like `ido.nvim/core` or should the core of Ido remain as `shoumodip/ido.nvim`?

**Comment by alexanderjeurissen** [comment]
I would move `shoumodip/ido.nvim` to the org as well just as `ido.nvim/ido.nvim` or as you suggested `ido.nvim/core`  it's a pattern that you see very often.

for instance for the neovim org: https://github.com/neovim/neovim
or the telescope.nvim package: https://github.com/nvim-telescope/telescope.nvim

I think it works better than keeping these separate as it becomes very scattered and hard to find complementary plugins that way. Examples include `ivy/swiper` in the emacs world where other related plugins are scattered all over the place which in my recollection when I used emacs made it very hard to find complementary plugins.

**Comment by shoumodip** [comment]
We first need to create a solid extensions interface. I think the following two features are essential.
- Automatic per-extension history management
- Fully customizable extensions, by means of a `ido.ext.load(EXT_NAME, OPTIONS)` or something

I was thinking instead of calling them extensions, why not call them `packages`. Ido is a framework, right? So it makes sense that stuff created using it should be called *packages* or *plugins*, not extensions.

Tell me if there are some other specifications we need for packages/extensions, because this needs to be part of Ido core before we create the `ido.nvim` organization. There's no point in creating an extensions organisation if the extension system in the core is lacking, am i right?

**Comment by alexanderjeurissen** [comment]
> We first need to create a solid extensions interface. I think the following two features are essential.
I'm not convinced that this is a blocker, I think with the current api in place `ido sources` / packages can already be created.
For example I myself already created several locally for things like:

- colorscheme
- project files (using silver searcher / rip grep)
- git files
- modified git files 

currently working on searching help tags.

All of these could already live as packages and users can simply install them like

```
Plug 'ido.nvim/git' 
```

and use them like

```lua

 IDO_GIT_OPTS = {
    prompt = "Git",
    layout = some_layout
  }

  set_keymap('n',
    '<leader>gf',
    "<cmd>lua require('ido.nvim/git').start(IDO_GIT_OPTS)<CR>",
    { noremap = true, silent = true })
```

in addition advices are not sandboxed and thus can be overwritten globally by extensions afaik. So that would open up things like packages that provide rendering enhancements (for instance https://github.com/ryanoasis/vim-devicons support in the file browser)

> 
> * Automatic per-extension history management
What history would these extensions have to maintain ?

> * Fully customizable extensions, by means of a `ido.ext.load(EXT_NAME, OPTIONS)` or something
> 
> I was thinking instead of calling them extensions, why not call them `packages`. Ido is a framework, right? So it makes sense that stuff created using it should be called _packages_ or _plugins_, not extensions.

I agree packages is a better term for this 👍 

> 
> Tell me if there are some other specifications we need for packages/extensions, because this needs to be part of Ido core before we create the `ido.nvim` organization. There's no point in creating an extensions organisation if the extension system in the core is lacking, am i right?

I think more granular advices, for instance I think a `item_renderer` advice where one can only overwrite the way each individual `item` is rendered would be *really* good. Currently one can only overwrite the complete `renderer` which gives a lot of power but also is not super practical for all use-cases. There might be more examples like this but this one is something I came across when browsing the code.

**Comment by shoumodip** [comment]
> I think with the current api in place ido sources / packages can already be created.

I don't mean we need to create a packages module or api. What I mean is to create a function which acts as a bit of syntactic sugar. For example the `api.options.setup()` function. It is not essential to use it, however it adds a little bit of beauty to your code.

So the code you suggested
```lua
IDO_GIT_OPTS = {
   prompt = "Git",
   layout = some_layout
}

set_keymap('n',
   '<leader>gf',
   "<cmd>lua require('ido.nvim/git').start(IDO_GIT_OPTS)<CR>",
   { noremap = true, silent = true })
```

could also be written as
```lua
ido.pkg.enable("git", {
   options = {
      prompt = "Git",
      layout = some_layout,
   },

   bind = {
      mode = "normal",
      key = "<Leader>gf",
      noremap = true,
      silent = true,
   },
})
```

Both options would be valid, but imo the second one *looks* better.

> What history would these extensions have to maintain ?

By history management I mean like in the shell.
- You accept an item
- It gets stored in the history
- You can go back and forth in history to get quick access to recent items

I personally think that each package should have its own history. LIke you don't want your `git` history of Ido intersecting with your `file_browser` history or whatever. So each package could be created like
```lua
local git = {}

function git.main()
   -- Package code
end

-- ido.pkg.create() is a function which merely acts as syntactic sugar. You may or may not use it.
ido.pkg.create("git", {

   -- The default options
   opts = {
      prompt = "Git",
      layout = some_layout,
   },

   -- By default bound to nothing
   bind = false,

   -- Entry point of the package, like the main() procedure in C
   main = git.main,
})

return git
```

and that should setup a table in the `git` module as the history like `git.history` or whatever.

> I think more granular advices, for instance I think a item_renderer advice where one can only overwrite the way each individual item is rendered would be really good.

By granulated rendering of items, do you mean the results or every single thing that is rendered like prompt, query, suggestion, etc?

**Comment by shoumodip** [comment]
To be fair, all of the functions in the `ido` module is syntactic sugar. So I don't think syntactic sugar for this would be bloat per-say.

**Comment by shoumodip** [comment]
Introduced the **optional** packages API. Check it out [here](https://github.com/shoumodip/ido.nvim/blob/main/wiki/packages.md).

> Use `ido.start()` when you wish to do something quick or you can't be bothered to use the packages API. Packages are optional. They merely provide an integrated and uniform experience when using custom Ido functions.

The package API is optional and doesn't interfere with the already established API. So your custom functions should work fine. Just change `options` to `opts` and `variables` to `vars`. I got tired of typing out these verbose variable names so I shortened them a bit in the code-base.

Tell me what you think of this. Is it bloat? Is it feature-creep? I am willing to listen to your opinion.
